lib/api/objectCopy.js

```diff
@@ -91,12 +93,20 @@ function _prepMetadata(sourceObjMD, headers, sourceIsDestination, authInfo,
         });
         return { error: userMetadata };
     }
-    // If metadataDirective is 'COPY' and location constraint header is
-    // specified, new location constraint should override source object's
-    if (whichMetadata === 'COPY' && headers[locationHeader]) {
-        userMetadata[locationHeader] = headers[locationHeader];
+    // If metadataDirective is:
+    // - 'COPY' and source object has a locaction constrain in its metadata
+    // we use the bucket destination location constraint
+    if (whichMetadata === 'COPY'
+        && userMetadata[locationHeader]
+        && destBucketMD.getLocationConstraint()) {
+        userMetadata[locationHeader] = destBucketMD.getLocationConstraint();
     }
-    const sourceLocationConstraintName = sourceObjMD[locationHeader];
+    const backendInfoObjSource = locationConstraintCheck(request,
+        sourceObjMD, sourceBucketMD, log);
+    if (backendInfoObjSource.err) {
+        return { error: backendInfoObjSource.err };
+    }
+    const sourceLocationConstraintName = backendInfoObjSource.controllingLC;
     // If location constraint header is not included, locations match
     const locationMatch = headers[locationHeader] ?
         sourceLocationConstraintName === headers[locationHeader] : true;
```

Comment:
shouldn't you be comparing the sourceLocationConstraintName to the destinationLocation (which could be the userMetadata[locationHeader] but might not be if you haven't set it? In any case, the headers seem to be the wrong place to look and we should always be checking against the actual destination.